virtual_rainforest/core/grid.py

```diff
@@ -1,203 +1,370 @@
+def register_grid(grid_type: str):
+    """Add a grid type and creator function to the grid registry.
 
+    This decorator is used to add a function creating a grid layout to the registry of
+    accepted grids. The function should return a dictionary of shapely.geometry.Polygon
+    objects, keyed by a cell id str.
 
-class CoreGrid:
-    """Define the grid of cells used in a virtual rainforest simulation.
+    The grid_type argument is used to identify the grid creation function to be used by
+    the Grid class and in configuration files.
 
     Args:
-        config: A CoreGridConfig instance used to setup the grid
-
-    Attrs:
-        config: The CoreGridConfig used to create an instance
-        cell_dict: A dictionary, keyed by cell id, of dictionaries, containing the cell
-            polygon, centroid and neighbours for each cell.
+        grid_type: A name to be used to identify the grid creation function.
     """
 
-    def __init__(self, config: CoreGridConfig) -> None:
+    def decorator_register_grid(func: Callable):
+
+        # Add the grid type to the registry
+        if grid_type in GRID_REGISTRY:
+            LOGGER.warning(
+                "Grid type %s already exists and is being replaced", grid_type
+            )
+        GRID_REGISTRY[grid_type] = func
+
+        # Pass the function through for use, using @wraps to expose the original
+        # function dunder attributes (__name__, __doc__ etc.)
+        @wraps(func)
+        def wrapper_register_grid(*args, **kwargs):
+
+            return func(*args, **kwargs)
+
+        return wrapper_register_grid
+
+    return decorator_register_grid
 
-        self.config = config
 
-        # Set up the grid
-        if config.grid_type == "square":
-            self.cell_dict = self._make_square_grid(
-                cell_area=config.cell_area,
-                cell_nx=config.cell_nx,
-                cell_ny=config.cell_ny,
+@register_grid(grid_type="square")
+def make_square_grid(
+    cell_area: float,
+    cell_nx: int,
+    cell_ny: int,
+    xoff: float = 0,
+    yoff: float = 0,
+) -> dict:
+    """Create a square grid.
+
+    Args:
+        cell_area: The area of each hexagon cell
+        cell_nx: The number of grid cells in the X direction.
+        cell_ny: The number of grid cells in the Y direction.
+        xoff: An offset to use for the grid origin in the X direction.
+        yoff: An offset to use for the grid origin in the Y direction.
+    """
+
+    # Create the polygon prototype object, with origin at 0,0 and area 1
+    # Note coordinate order is anti-clockwise - right hand rule.
+    prototype = Polygon([[0, 0], [1, 0], [1, 1], [0, 1], [0, 0]])
+
+    # Scale to requested size and origin
+    scale_factor = np.sqrt(cell_area)
+    prototype = scale(prototype, xfact=scale_factor, yfact=scale_factor, origin=(0, 0))
+    prototype = translate(prototype, xoff=xoff, yoff=yoff)
+
+    # Tile the prototypes to create the grid
+    cell_dict = {}
```

Comment:
I wonder if a dict is the best option here (no idea - it depends on the end use case). Personally, I would use a nested list - given that numpy arrays won't work in this case - so I can do simple indexing like `cell[0][3]` without having to create a string key first. Probably a pandas DataFrame would work, so you can access independently whole rows and columns.

But again, it depends on the use case.


```

Reply:
I'm going to leave it as is for the moment. We need to think about how we structure spatial data internally, and I think there might be good reasons to flatten X and Y and just make this a flat list with a single integer index for location across cells. I agree that using numeric indices and not string keys is cleaner, but I'm going to wait until we have a decision there.